### ISSUE TITLE
support unix/dotnet style switches

### ISSUE THREAD

**Issue by danmoseley** [dotnet/msbuild#7977]
Probably there's existing discussion/ a dupe that I didn't find. https://github.com/dotnet/msbuild/issues/5714 seems somewhat related.

MSBuild switch syntax was not created with Unix ways in mind, and this has become more obvious in recent years on Unix, and also on Windows when using tools like dotnet.exe on Windows that prefer Unix conventions and aggregate MSBuild switches. For example. `build.cmd -ci -arch x64 -os windows -s clr.aot+libs -rc Release -lc Release  /p:CrossBuild=false ` (an example of a command to build dotnet/runtime) and `dotnet build -c debug /t:test`

Ideally I could write `build.cmd -ci -arch x64 -os windows -s clr.aot+libs -rc Release -lc Release  -p CrossBuild=false ` and `dotnet build -c debug -t test`

MSBuild already accepts `--XXX` for multi character switches and `-X` for single character ones. It also does not reject `--` but does not respect it either -- for example `nonexistent.csproj --` produces "Project file does not exist" even though I'd expect it to look like an invalid flag rather than a file. To be fully consistent it would need to (I think):

1. Allow "valued" switches to use space rather than colon
2. Ideally respect "--" 

I'm wondering whether this could be possible without generally breaking existing scripts. If I have `dotnet build -c debug -t:test -p:configuration=debug my.csproj` and now write `dotnet build -c debug -t test -p configuration=debug my.csproj`, it should still be possible to unambiguously that the project was `my.csproj` and not `test`. However, optional-valued switches may be a problem. If `-m:4` becomes `-m 4`, is the 4 the project name or should it continue to implicitly build a project in the working directory?

Just curious whether there's been discussion about this, as the mixture of colon and space style is a bit ugly.

**Comment by rainersigwald** [comment]
I bet @baronfel will have thoughts here.

What would you want `--` to indicate? I don't see an analogous MSBuild operation to "everything from here is a single string you shouldn't parse".

**Comment by rainersigwald** [comment]
> MSBuild already accepts `--XXX` for multi character switches and `-X` for single character ones.

We actually respect `-` `--` and `/` prefixes for all of our arguments today, regardless of single/multichar.

**Comment by danmoseley** [comment]
> What would you want -- to indicate?

It would mean "nothing before here is a project file, everything after is a project file". I'm not sure whether it would be needed to break ambiguity unless a file name started with -- so perhaps it's minor/low priority?

> We actually respect - -- and / prefixes for all of our arguments today, regardless of single/multichar.

Yup -- I think it'd be a clear non goal to reject syntax that is not Unix. Just to allow Unix syntax.

**Comment by baronfel** [comment]
If we look to add unix support, we should _really really_ look into just evaluating if System.CommandLine can represent the MSBuild CLI. It already handles most (all?) of the questions described above (regarding precedence of `-m` for example) in an unambiguous way. It also supports `--` in the UNIX way by default, meaning it parses each of the tokens to the right as an 'argument' and allows the application to decide what to do with them (which mostly means forwarding along, but can mean many other things).

**Comment by danmoseley** [comment]
Does MSBuild parse arguments outside of the context of 'dotnet' ? As I loosely understand it, many of the sub-commands of 'dotnet' already use System.CommandLine, in some kind of delegated way.
cc @terrajobst as we were talking about this

**Comment by baronfel** [comment]
@danmoseley MSBuild has a parser of its own and it actually runs all the time even if the .NET CLI calls into MSBuild. The .NET CLI calls the `main` method of MSbuild directly with a string[] if ran in-proc, otherwise it's shelling out to `dotnet msbuild.dll` with those same arguments.

Separately, msbuild.exe for full framework needs to parse as well, so any .NET CLI-focused solution will pass that binary by entirely.

**Comment by KathleenDollard** [comment]
For anyone that does not know , the main part of the .NET CLI uses System.CommandLine but still in a non-standard way because it has a lot of legacy code to clean up. @baronfel has led cleaning that up on a sane timeline. `dotnet new` uses System.CommandLine starting in .NET 7, and it was key to modernizing the Ux. 

Also, @vlada-shubina did great work to supply hidden backwards compat for `dotnet new` to avoid breaking scripts. I do not know if that would be possible for MSBuild.

**Comment by danmoseley** [comment]
As an aside, I've sometimes wanted to pass more than one project (to build in parallel) so I hope any parser would not assume there was one or zero.

**Comment by rainersigwald** [comment]
> > It would mean "nothing before here is a project file, everything after is a project file".
> 
> According to [Utility Syntax Guidelines](https://pubs.opengroup.org/onlinepubs/009695399/basedefs/xbd_chap12.html#tag_12_02) in _The Open Group Base Specifications Issue 6_, the arguments that follow `--` can only be operands, but the arguments that precede `--` can be options or operands.

> Git deviates from this by not allowing file name operands before `--`.

That doesn't appear to be true to me:

```sh-session
❯ git rm .\global.json -- .\NuGet.config
rm 'NuGet.config'
rm 'global.json'
```

> I don't think MSBuild should follow Git's example.

We would not take any breaking changes to the argument parsing, only augmentations.

> I've sometimes wanted to pass more than one project (to build in parallel) so I hope any parser would not assume there was one or zero.

This has definitely come up before but I can't find a bug. It's a good idea IMO but doesn't need to be tied to this change.

**Comment by danmoseley** [comment]
> This has definitely come up before but I can't find a bug. It's a good idea IMO but doesn't need to be tied to this change.

Right, I just mean that I think it could be added if we choose at some point without introducing ambiguity, I hope that remains true

**Comment by danmoseley** [comment]
> That doesn't appear to be true to me:

Perhaps it ignores it unless ambiguity exists.
```
C:\git\runtime>dir > head

C:\git\runtime>git show -q HEAD
fatal: ambiguous argument 'HEAD': both revision and filename
Use '--' to separate paths from revisions, like this:
'git <command> [<revision>...] -- [<file>...]'

C:\git\runtime>del HEAD

C:\git\runtime>git show -q HEAD
commit 1d025196ae0d0f1e0b17e2e6033234e1937d997b (HEAD -> main, upstream/main, upstream/HEAD)
```
